nav2_smac_planner/include/nav2_smac_planner/goal_manager.hpp

```diff
@@ -0,0 +1,194 @@
+* @brief Responsible for managing multiple variables storing information on the goal
+*/
+template<typename NodeT>
+class GoalManager{
+public:
+  typedef NodeT * NodePtr;
+  typedef std::vector<NodePtr> NodeVector;
+  typedef std::unordered_set<NodePtr> NodeSet;
+  typedef std::vector<GoalState<NodeT>> GoalStateVector;
+  typedef typename NodeT::Coordinates Coordinates;
+  typedef typename NodeT::CoordinateVector CoordinateVector;
+
+   /**
+   * @brief Constructor: Initializes empty goal state. sets and coordinate lists.
+   */
+  GoalManager()
+  : _goals_set(NodeSet()),
+    _goals_state(GoalStateVector()),
+    _goals_coordinate(CoordinateVector())
+  {
+  }
+
+  /**
+   * @brief Destructor for the GoalManager
+   */
+  ~GoalManager()
+  {
+  }
+
+  /**
+   * @brief Checks if the goals set is empty
+   * @return true if the goals set is empty
+   */
+  bool goalsIsEmpty()
+  {
+    return _goals_state.empty();
+  }
+
+  /**
+   * @brief Sets the internal goals' state list to the provided goals.
+   * @param goals Vector of goals state to set,
+   */
+
+  void setGoalStates(
+    GoalStateVector & goals_state)
+  {
+    clear();
+    _goals_state = goals_state;
+  }
+
+  /**
+   * @brief Clears all internal goal data, including goals, states, and coordinates.
+   */
+  void clear()
+  {
+    _goals_set.clear();
+    _goals_state.clear();
+    _goals_coordinate.clear();
+  }
+
+  /**
+   * @brief Populates coarse and fine goal lists for analytic expansion.
+   * @param coarse_check_goals Output list of goals for coarse search expansion.
+   * @param fine_check_goals Output list of goals for fine search refinement.
+   * @param coarse_search_resolution Number of fine goals per coarse goal.
+   */
+  void prepareGoalsForAnalyticExpansion(
+    NodeVector & coarse_check_goals, NodeVector & fine_check_goals,
+    int coarse_search_resolution)
+  {
+    for (unsigned int i = 0; i < _goals_state.size(); i++) {
+      if (_goals_state[i].is_valid) {
+        if (i % coarse_search_resolution == 0) {
+          coarse_check_goals.push_back(_goals_state[i].goal);
+        } else {
+          fine_check_goals.push_back(_goals_state[i].goal);
+        }
+      }
+    }
+  }
+
+  /**
+   * @brief Filters and marks invalid goals based on collision checking and tolerance thresholds.
+   *
+   * Stores only valid (or tolerably infeasible) goals into internal goal sets and coordinates.
+   *
+   * @param tolerance Heuristic tolerance allowed for infeasible goals.
+   * @param collision_checker Collision checker to validate goal positions.
+   * @param traverse_unknown Flag whether traversal through unknown space is allowed.
+   */
+  void removeInvalidGoals(
+    const float & tolerance,
+    GridCollisionChecker * collision_checker,
+    const bool & traverse_unknown)
+  {
+    const bool allow_infeasible = tolerance > 0.001;
+       // Predicate: should we try this goal?
+    auto shouldTryGoal = [&](const NodePtr & node_ptr) {
```

Comment:
Is the lambda still necessary?

Reply:
actually no, just looked cleaner for me but i can remove it.

Reply:
done